Preliminary DeviceMesh + native c10d functional integration

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118498
* #118437
* __->__ #118423

### Summary
- Added `group_name` as the third field in `dim_group_infos`.
- `DeviceMeshTest` now runs both w/ and w/0 `_USE_NATIVE_C10D_FUNCTIONAL=1` in CI.

### Other fixes
- Convert `reduceOp` to lower case before passing it into c10d_functional ops.
- Added a finalizer to handle unwaited collectives (this mirrors the treatment for Python functional collective ops).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
Preliminary DeviceMesh + native c10d functional integration

[ghstack-poisoned]

wanchaol commented on torch/distributed/device_mesh.py:
```diff
@@ -254,9 +254,13 @@ def _get_or_create_default_group(self):
+            # TODO(yifu): remove tag and ranks once we fully migrate to native
+            # functional collectives. See details in:
+            # https://github.com/pytorch/pytorch/issues/93173#issuecomment-1907095208
+            dim_group_infos: List[Tuple[str, List[int], str]] = []
```
quick comment: would `tag` have the same semantic with `group_name`? i.e. are they the same str? If they are maybe we can reuse the tag?

yifuwang replied:
They are different things. tag isn't always present so it's tricky to reuse.

Commit:
Update on "Preliminary DeviceMesh + native c10d functional integration"


### Summary
- Added `group_name` as the third field in `dim_group_infos`.
- Without `_USE_NATIVE_C10D_FUNCTIONAL=1`, all tests in `DeviceMeshTest` pass.
- With `_USE_NATIVE_C10D_FUNCTIONAL=1`, `DeviceMeshTest::test_fake_pg_device_mesh` fails. This is expected since native functional collective ops requires backends to be implemented in C++. I will port `FakeProcessGroup` to C++ in the next PR.
  - Note that the test failures do not affect CI since `_USE_NATIVE_C10D_FUNCTIONAL` is not set by default.

### Other fixes
- Convert `reduceOp` to lower case before passing it into c10d_functional ops.
- Added a finalizer to handle unwaited collectives (this mirrors the treatment for Python functional collective ops).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented:
Suggest you land the C++ FakePG first (restack for a different landing order) and then ensure that when you land the native_c10d functional ops, both 0 and 1 paths are both CI tested from first place

Commit:
Update on "Preliminary DeviceMesh + native c10d functional integration"


### Summary
- Added `group_name` as the third field in `dim_group_infos`.
- Without `_USE_NATIVE_C10D_FUNCTIONAL=1`, all tests in `DeviceMeshTest` pass.
- With `_USE_NATIVE_C10D_FUNCTIONAL=1`, `DeviceMeshTest::test_fake_pg_device_mesh` fails. This is expected since native functional collective ops requires backends to be implemented in C++. I will port `FakeProcessGroup` to C++ in the next PR.
  - Note that the test failures do not affect CI since `_USE_NATIVE_C10D_FUNCTIONAL` is not set by default.

### Other fixes
- Convert `reduceOp` to lower case before passing it into c10d_functional ops.
- Added a finalizer to handle unwaited collectives (this mirrors the treatment for Python functional collective ops).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

yifuwang commented:
@wconstab reordered the PRs as you suggested.

wanchaol reviewed (approved):
Nice, sgtm!

LucasLLC reviewed (approved):
lgtm, tagging @wz337 since she is the device mesh expert

Commit:
Update on "Preliminary DeviceMesh + native c10d functional integration"


### Summary
- Added `group_name` as the third field in `dim_group_infos`.
- Without `_USE_NATIVE_C10D_FUNCTIONAL=1`, all tests in `DeviceMeshTest` pass.
- With `_USE_NATIVE_C10D_FUNCTIONAL=1`, `DeviceMeshTest::test_fake_pg_device_mesh` fails. This is expected since native functional collective ops requires backends to be implemented in C++. I will port `FakeProcessGroup` to C++ in the next PR.
  - Note that the test failures do not affect CI since `_USE_NATIVE_C10D_FUNCTIONAL` is not set by default.

### Other fixes
- Convert `reduceOp` to lower case before passing it into c10d_functional ops.
- Added a finalizer to handle unwaited collectives (this mirrors the treatment for Python functional collective ops).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab reviewed (approved):
cool lgtm! (maybe update PR desc, still says tests fail bc of fake PG)

yifuwang commented:
@pytorchbot merge

yifuwang commented:
@pytorchbot merge
